app/utils/smartapi/validator.py

```diff
@@ -43,3 +60,148 @@ def validate_symbol_and_get_token(
+    -----------
+    InvalidDateTimeFormatException:
+        If the date and time is in wrong format.
+    Return:
+    -------
+    datetime
+        validated given date and time and returns valid datetime object.
+    """
+    date_separator = "/" if len(date_time.split("/")) > 1 else "-"
+    if re.search("[a-zA-Z]+", date_time):
+        month_format_codes = ["b", "B"]
+    else:
+        month_format_codes = ["m", "-m"]
+    day_format_codes = ["d", "-d"]
+    hour_format_codes = ["H", "-H"]
+    minute_format_codes = ["M", "-M"]
+    for month_format_code in month_format_codes:
+        for day_format_code in day_format_codes:
+            for hour_format_code in hour_format_codes:
+                for minute_format_code in minute_format_codes:
+                    try:
+                        datetime_format = (
+                            f"%Y{date_separator}%{month_format_code}{date_separator}"
+                            f"%{day_format_code} %{hour_format_code}:%{minute_format_code}"
+                        )
+                        datetime_obj = datetime.strptime(date_time, datetime_format)
+                        return datetime_obj
+
+                    except ValueError:
+                        continue
+    raise InvalidDateTimeFormatException(date_time)
+
+
+def check_market_open_between_dates(start_date: datetime, end_date: datetime):
+    """Checks whether the stock market is open between the given dates.
+
+    Parameters:
+    -----------
+    start_date: `datetime`
+        Start date.
+    end_date: `datetime`
+        End date.
+
+    Exceptions:
+    -----------
+    AllDaysHolidayException:
+        Raised when all days in the given date range are market holidays.
+    """
+    holidays_data = read_text_data(NSE_HOLIDAYS_PATH)
+    current_date = start_date
+    while current_date <= end_date:
+        index = bisect_left(holidays_data, current_date.strftime("%Y-%m-%d"))
+        if (
+            current_date.weekday() < 5
+            and index != len(holidays_data)
+            and holidays_data[index] != current_date.strftime("%Y-%m-%d")
+        ):
+            return
+        current_date += timedelta(days=1)
+    raise AllDaysHolidayException(
+        start_date.strftime("%Y-%m-%d"), end_date.strftime("%Y-%m-%d")
+    )
+
+
+def check_data_availability(end_date: datetime, stock_symbol: str):
+    """Check whether the data is available or not for the requested dates from the SmartAPI.
+
+    Parameters:
+    -----------
+    end_date: `datetime`
+        End date
+    stock_symbol: `str`
+        Symbol of the stock.
+
+    Exceptions:
+    -----------
+    DataUnavailableException:
+        Raised when the data is unavailable for the requested dates from the SmartAPI.
+    """
+    data_starting_dates = load_json_data(DATA_STARTING_DATES_PATH)
+    starting_date = data_starting_dates.get(stock_symbol)
+    if starting_date is None or end_date < datetime.strptime(starting_date, "%Y-%m-%d"):
```

Comment:
because if the end date less than the starting date then there is no way to get data due to we only have data from starting date onwards. if end date is greater than the starting date then there must be at least one day data which will not be empty.